### PR TITLE
[Record] 3-Layer Depth Recurrence + EMA 0.9965 + WD 0.095 — val_bpb 1.0889

### PR DESCRIPTION
**Pull request by X-Abhishek-X**
## Record: 3-Layer Depth Recurrence + EMA 0.9965 + WD 0.095 — val_bpb 1.0889

**val_bpb: 1.0889** (3-seed mean, std 0.0005) | **~15.89 MB** | 8×H100 SXM, 590s

### 3-Seed Results (8×H100 80GB SXM)

| Seed | Pre-quant BPB | **Sliding BPB (s64)** | Artifact |
|------|---------------|----------------------|----------|
| 42 | 1.0950 | **1.0885** | 15,890,417 B |
| 1337 | 1.0959 | **1.0894** | 15,888,733 B |
| 2024 | 1.0954 | **1.0888** | 15,895,711 B |
| **Mean** | **1.0954** | **1.0889 (std 0.0005)** | |

Current merged SOTA: **1.1147** (PR #1019). Delta: **−0.0258 BPB**.

### Key Changes

Four refinements stacked on PR #1334's depth recurrence architecture:

| Parameter | PR #1334 | This | Source |
|-----------|----------|------|--------|
| **Recurrence layers** | 4,5 (2-layer) | **3,4,5 (3-layer)** | PR #1331 |
| **Weight decay** | 0.090 | **0.095** | PR #1331 |
| **Matrix LR** | 0.020 | **0.022** | PR #1331 |
| **EMA decay** | 0.997 | **0.9965** | PR #1421 (this author) |
| **Recurrence start** | step 3000 | **step 2000** | This work |
| **Warmdown fraction** | 0.667 | **0.72** | This work |

### Why This Combination Works

- **3-layer recurrence (3,4,5)**: 14 virtual layers from 11 physical. More compute per forward pass without additional parameters.
- **WD=0.095 + MLR=0.022**: Higher WD compresses weights, improving GPTQ quantization. Only 134K-186K values pruned.
- **EMA decay=0.9965**: Smoother weight averaging for cleaner quantization.
- **Early recurrence (step 2000)**: 1000 more training steps with full depth recurrence.
- **Extended warmdown (72%)**: Weights fully settle before GPTQ.

### Architecture (from PR #1334)

- 11 transformer layers, 512-dim, 8 heads (4 KV heads, GQA)
- Depth recurrence: layers 3,4,5 repeat (virtual 14 layers), activated at step 2000
- Skip gates, parallel residuals from layer 7, QK-Gain 5.0
- Shared Value Embedding (dim=128, layers 9,10)
- Tied embeddings, logit softcap=30.0, SP4096 tokenizer

### Training

- FlashAttention 3, Muon (lr=0.022, WD=0.095), Adam/AdamW (fused=True)
- Gradient clip: 0.3, Batch: 786,432 tokens/step, seq_len=2048
- Warmdown: 72%, EMA decay=0.9965, Wallclock: 590s

### Quantization

- GPTQ int6, percdamp=0.05, 64 calibration batches
- Selective pruning (~134K-186K values), Brotli compression

### Run Command

```bash
SEED=42 RECUR_START_STEP=2000 WARMDOWN_FRAC=0.72 \
torchrun --standalone --nproc_per_node=8 train_gpt.py
```

### Credits

- **Base architecture + depth recurrence**: PR #1334 by @aryanbhosale
- **3-layer recurrence + WD/LR tuning**: PR #1331
- **EMA decay tuning (0.9965)**: PR #1421 by @X-Abhishek-X (this author)
- **Early recurrence + extended warmdown**: This work

### CODE REVIEW

**Comment by MatoTeziTanka** [comment]
### Community Review — [Record] 3-Layer Depth Recurrence + EMA 0.9965 + WD 0.095 — val_bpb 1.0889

**BPB:** 1.0889 | **Compliance:** LOOKS CLEAN — score-first-per-chunk TTT (legal #1416/#1423 pattern)

**What I found in the code** (head SHA `be838e4a502b`, file `records/track_10min_16mb/2026-04-07_3Layer_DepthRecurrence_EMA0.9965_WD095_1.0889/train_gpt.py`):

The TTT path at line **1521** implements the score-first-per-chunk pattern: each chunk is scored under `torch.no_grad()` / `inference_mode()` before the `base_model.train()` + SGD adaptation runs on that same chunk, with an `is_last_chunk` guard so the final chunk gets no adaptation pass. This is the structural shape the legal frontier uses (PRs #1416 erichroepke, #1423 aryanbhosale).

Per **Issue #402 and Issue #677**, TTT is legal when each token is scored before the adapter updates on it, and that's what the code does here — chunk `ci` is scored under weights adapted only on chunks `0..ci-1`. No `prequant_ttt_adapt_adamw(val_tokens, ...)` multi-epoch fine-tune, no scored-region SLOT, no target-in-key n-gram cache.

**CPU smoke test (CT2038 proteus-engine, 2026-04-11):** import OK in 14.62s, dim=512, layers=11, vocab=4096, code=83569 B, SMOKE_TEST_PASS

**Verdict:** LOOKS CLEAN.

**Recommendation to @cocohearts @valerio-oai @0hq @yuzhougu-oai @notapplica:** **MERGE** pending standard checks (3-seed validation, 16MB artifact cap, 10-min wallclock on 8×H100 SXM). The compliance picture matches the legal reference frontier and no flags were raised by the classification pass.

*Auto-classification caveat:* this review was drafted by the AST-based classifier against a template derived from manually-reviewed cluster PRs (#1420, #1450, #1487, #1541, #1529, #1533, #1518). If I've misread a subtlety in your eval path — e.g., multi-epoch TTT that I mistook for single-pass, or a target-in-key lookup I missed in a helper function — please flag it and I'll re-run the audit manually.

---
*Reviewed by [@MatoTeziTanka](https://github.com/MatoTeziTanka) — [The Agora](https://matotezitanka.github.io/parameter-golf). CPU smoke test (CT2038 proteus-engine, 2026-04-11): import OK in 14.62s, dim=512, layers=11, vocab=4096, code=83569 B, SMOKE_TEST_PASS. Classification via deterministic AST-based `classify_prs.py` (pattern bank derived from ~65 manually-reviewed PRs earlier in the 2026-04-11 sweep). This review was auto-drafted from a template and spot-checked before posting — if the template misread your code, please call it out so I can iterate the classifier.*
